File: 01 clase entero/js/enteros.js
class Entero {
    // Atributos
    Num;

    // Constructor
    constructor(Numero){
        this.Num = Numero;
    }

    // CLASS - Cargar el valor de Num
    setNum(){
        const input = document.getElementById('numberInput');
        this.Num = parseInt(input.value, 10);
    }

    // CLASS - Obtener el valor de Num
    getNum() {
        return this.Num;
    }

    // CLASS - Mostrar el valor de Num
    showNum() {
        const resultadoDiv = document.getElementById('resultado');
        resultadoDiv.textContent = this.getNum();
    }

    // Obtener el valor de Num y mostrarlo
    showResultado(respuesta) {
        const resultadoDiv = document.getElementById('resultado');
        resultadoDiv.textContent = respuesta;
    }

    incrementarNum(){
        this.Num = this.Num + 1;
    }

    decrementarNum(){
        this.Num = this.Num - 1;
    }

    esParImpar(){
        return this.Num % 2 === 0;
    }

    esNegativoPositivo(){
        return this.Num >= 0;
    }

    factorial(){
        if(this.Num < 0) {
            return "No existe factorial de números negativos";
        }
        let resultado = 1;
        for (let i = 1; i <= this.Num; i++) {
            resultado *= i;
        }
        return resultado;
    }

    // VERIFICAR SI ES PERFECTO
    esPerfecto(){
        let sumaDivisores = 0;
        for (let i = 1; i < this.Num; i++) {
            if (this.Num % i === 0) {
                sumaDivisores += i;
            }
        }
        return sumaDivisores === this.Num;
    }

    // VERIFICAR SI ES PRIMO
    esPrimo() {
        if (this.Num <= 1) return false; 
        if (this.Num === 2) return true; 
        if (this.Num % 2 === 0) return false; 
        
        for (let i = 3; i <= Math.sqrt(this.Num); i += 2) {
            if (this.Num % i === 0) return false;
        }
        return true;
    }

    // CALCULAR SUCESIÓN DE FIBONACCI
    fibonacci() {
        if (this.Num <= 0) return "Número debe ser positivo";

        let serie = [];
        let a = 0;
        let b = 1;

        for (let i = 0; i < this.Num; i++) {
            serie.push(a);
            let temp = a + b;
            a = b;
            b = temp;
        }

        return serie.join(', ');
    }

    esArmstrong() {
        let numStr = this.Num.toString();  // Convertimos el número a cadena
        let numLength = numStr.length;  // Obtenemos el número de dígitos
        let suma = 0;
        // Sumamos los cubos de los dígitos
        for (let i = 0; i < numLength; i++) {
            suma += Math.pow(parseInt(numStr[i]), 3);
        }
        // Comparamos si la suma de los cubos es igual al número original
        return suma === this.Num;
    }

    esPerfectoGeneralizado(k) {
        let suma = 0;
        // Recorremos los divisores propios del número
        for (let i = 1; i < this.Num; i++) {
            if (this.Num % i === 0) {
                // Sumamos la k-ésima potencia de los divisores propios
                suma += Math.pow(i, k);
            }
        }
        // Comparamos la suma con la k-ésima potencia del número original
        return suma === Math.pow(this.Num, k);
    }

    calcularFilaPascal() {
        let fila = [1]; // La primera posición siempre es 1
        for (let i = 1; i <= this.Num; i++) {
            fila[i] = fila[i - 1] * (this.Num - i + 1) / i;
        }
        return fila;
    }

    cifrarCesar(desplazamiento) {
        let resultado = '';
        let texto = this.Num.toString(); // Convertir el número a cadena de texto si es un número
        for (let i = 0; i < texto.length; i++) {
            let char = texto[i];
            if (char.match(/[a-z]/i)) { // Verifica si el carácter es una letra
                let codigo = char.charCodeAt(0);
                // Manejo de mayúsculas
                if (codigo >= 65 && codigo <= 90) {
                    char = String.fromCharCode(((codigo - 65 + desplazamiento) % 26) + 65);
                }
                // Manejo de minúsculas
                else if (codigo >= 97 && codigo <= 122) {
                    char = String.fromCharCode(((codigo - 97 + desplazamiento) % 26) + 97);
                }
            }
            resultado += char; // Añade el carácter cifrado al resultado
        }
        return resultado;
    }
    


    tienePrimoGemelo() {
        // Función auxiliar para verificar si un número es primo
        const esPrimo = (num) => {
            if (num <= 1) return false;
            for (let i = 2; i <= Math.sqrt(num); i++) {
                if (num % i === 0) return false;
            }
            return true;
        };
        // Verificar si N tiene un primo gemelo
        if (esPrimo(this.Num)) {
            // Verificar si N-2 o N+2 también son primos
            return esPrimo(this.Num - 2) || esPrimo(this.Num + 2);
        }
        return false;
    }

    descomponerEnFactoresPrimos() {
        let num = this.Num;
        let factores = [];
        // Encontrar el número de veces que 2 divide a num
        let potencia = 0;
        while (num % 2 === 0) {
            num /= 2;
            potencia++;
        }
        if (potencia > 0) factores.push([2, potencia]); // Añadir el factor 2 con su potencia
        // Encontrar otros factores primos impares
        for (let i = 3; i <= Math.sqrt(num); i += 2) {
            potencia = 0;
            while (num % i === 0) {
                num /= i;
                potencia++;
            }
            if (potencia > 0) factores.push([i, potencia]); // Añadir el factor primo con su potencia
        }
        // Si num es mayor que 2, es un factor primo restante
        if (num > 2) factores.push([num, 1]);
    
        return factores;
    }

    calcularCatalan() {
        // Función auxiliar para calcular el factorial
        const factorial = (num) => {
            let resultado = 1;
            for (let i = 1; i <= num; i++) {
                resultado *= i;
            }
            return resultado;
        };
        // Fórmula de los números de Catalan: C(n) = (2n)! / ((n+1)! * n!)
        let n = this.Num;
        let catalan = factorial(2 * n) / (factorial(n + 1) * factorial(n));
        
        return catalan;
    }
    esMersenne() {
        let num = this.Num;
        // Comprobamos si num + 1 es una potencia de 2
        let p = Math.log2(num + 1); 
        if (p % 1 !== 0) return false; // Si p no es entero, no es de Mersenne
        // Verificamos si p es primo
        return this.esPrimo(p); // Usamos la función de primo existente
    }

}

// Las funciones = button HTML
var ClaseEntero = new Entero(0); // Se inicializa en 0

// Crear Objeto
function cargarNum() {
    ClaseEntero.setNum();
}

function mostrarNum() {
    ClaseEntero.showNum();
}

// Incrementar el valor de Num
function IncrementarValor(){
    ClaseEntero.incrementarNum();
    ClaseEntero.showNum();
}

// Decrementar el valor de Num
function DecrementarValor(){
    ClaseEntero.decrementarNum();
    ClaseEntero.showNum();
}

// Verifica si es Par o Impar
function esParImparNum(){
    var respuesta = ClaseEntero.esParImpar();
    var resp = respuesta ? "Es un número par" : "Es un número impar";
    ClaseEntero.showResultado(resp);
}

function esPositivoNegativoNum(){
    var respuesta = ClaseEntero.esNegativoPositivo();
    var resp = respuesta ? "Es un número positivo" : "Es un número negativo";
    ClaseEntero.showResultado(resp);
}

// Mostrar el factorial
function CalcularFactorial(){
    var resultado = ClaseEntero.factorial();
    ClaseEntero.showResultado("Factorial: " + resultado);
}

// Verifica si es número perfecto
function verificarPerfecto(){
    var resultado = ClaseEntero.esPerfecto();
    var resp = resultado ? "Es un número perfecto" : "No es un número perfecto";
    ClaseEntero.showResultado(resp);
}

function verificarPrimo(){
    var resultado = ClaseEntero.esPrimo();
    var resp = resultado ? "Es un número primo" : "No es un número primo";
    ClaseEntero.showResultado(resp);
}

function mostrarFibonacci() {
    var resultado = ClaseEntero.fibonacci();
    ClaseEntero.showResultado("Serie de Fibonacci: " + resultado);
}
function esArmstrongNum(){
    var respuesta = ClaseEntero.esArmstrong();
    var resp = respuesta ? "Es Num Armstrong" : "No es Num Armstrong";
    ClaseEntero.showResultado(resp);
}
function esPerfectoGeneralizadoNum() {
    const k = parseInt(document.getElementById('kInput').value, 10);
    var respuesta = ClaseEntero.esPerfectoGeneralizado(k);
    
    // Usa comillas y el operador de concatenación correctamente
    var resp = respuesta 
        ? `Es Num perfecto generalizado con k=${k}` 
        : `No es Num perfecto generalizado con k=${k}`;
    
    ClaseEntero.showResultado(resp);
}


function calcularFilaPascalNum() {
    var respuesta = ClaseEntero.calcularFilaPascal();
    ClaseEntero.showResultado("Fila de Pascal: " + respuesta.join(", "));
}

function cifrarCesarNum() {
    var desplazamiento = parseInt(prompt("Ingresa el valor del desplazamiento:"));
    var texto = prompt("Ingresa el texto a cifrar:");
    ClaseEntero.Num = texto; // Asigna el texto ingresado a la propiedad Num
    var respuesta = ClaseEntero.cifrarCesar(desplazamiento);
    ClaseEntero.showResultado("Cifrado César: " + respuesta);
}


function tienePrimoGemeloNum() {
    var respuesta = ClaseEntero.tienePrimoGemelo();
    var resp = respuesta ? "Tiene un primo gemelo" : "No tiene primo gemelo";
    ClaseEntero.showResultado(resp);
}

function descomponerEnFactoresPrimosNum() {
    var respuesta = ClaseEntero.descomponerEnFactoresPrimos();
    var resultadoFormateado = respuesta.map(f => f[0] + "^" + f[1]).join(", ");
    ClaseEntero.showResultado("Factores primos: " + resultadoFormateado);
}

function calcularCatalanNum() {
    var respuesta = ClaseEntero.calcularCatalan();
    ClaseEntero.showResultado("Número de Catalan: " + respuesta);
}

function esMersenneNum() {
    var respuesta = ClaseEntero.esMersenne();
    var resp = respuesta ? "Es un número de Mersenne" : "No es un número de Mersenne";
    ClaseEntero.showResultado(resp);
}

File: 01 clase entero/py/enteros.py
class Entero:
    def __init__(self, numero=0):
        self.Num = numero

    def set_num(self, numero):
        self.Num = numero

    def get_num(self):
        return self.Num

    def incrementar_num(self):
        self.Num += 1

    def decrementar_num(self):
        self.Num -= 1

    def es_par_impar(self):
        return self.Num % 2 == 0

    def es_negativo_positivo(self):
        return self.Num >= 0

    def factorial(self):
        if self.Num < 0:
            return "No existe factorial de números negativos"
        resultado = 1
        for i in range(1, self.Num + 1):
            resultado *= i
        return resultado

    def es_perfecto(self):
        suma_divisores = sum(i for i in range(1, self.Num) if self.Num % i == 0)
        return suma_divisores == self.Num

    def es_primo(self):
        if self.Num <= 1:
            return False
        if self.Num == 2:
            return True
        if self.Num % 2 == 0:
            return False
        for i in range(3, int(self.Num ** 0.5) + 1, 2):
            if self.Num % i == 0:
                return False
        return True

    def fibonacci(self):
        if self.Num <= 0:
            return "Número debe ser positivo"
        serie = []
        a, b = 0, 1
        for _ in range(self.Num):
            serie.append(a)
            a, b = b, a + b
        return serie

    def es_armstrong(self):
        num_str = str(self.Num)
        suma = sum(int(d)**3 for d in num_str)
        return suma == self.Num

    def es_perfecto_generalizado(self, k):
        suma = sum(i**k for i in range(1, self.Num) if self.Num % i == 0)
        return suma == self.Num ** k

    def calcular_fila_pascal(self):
        fila = [1]
        for i in range(1, self.Num + 1):
            fila.append(fila[-1] * (self.Num - i + 1) // i)
        return fila

    def cifrar_cesar(self, desplazamiento, texto):
        resultado = ''
        for char in texto:
            if char.isalpha():
                codigo = ord(char)
                if char.isupper():
                    resultado += chr((codigo - 65 + desplazamiento) % 26 + 65)
                else:
                    resultado += chr((codigo - 97 + desplazamiento) % 26 + 97)
            else:
                resultado += char
        return resultado

    def tiene_primo_gemelo(self):
        def es_primo(num):
            if num <= 1:
                return False
            for i in range(2, int(num ** 0.5) + 1):
                if num % i == 0:
                    return False
            return True

        return es_primo(self.Num) and (es_primo(self.Num - 2) or es_primo(self.Num + 2))

    def descomponer_en_factores_primos(self):
        num = self.Num
        factores = []
        potencia = 0

        while num % 2 == 0:
            num //= 2
            potencia += 1
        if potencia > 0:
            factores.append((2, potencia))

        for i in range(3, int(num ** 0.5) + 1, 2):
            potencia = 0
            while num % i == 0:
                num //= i
                potencia += 1
            if potencia > 0:
                factores.append((i, potencia))

        if num > 2:
            factores.append((num, 1))
        return factores

    def calcular_catalan(self):
        def factorial(n):
            return 1 if n == 0 else n * factorial(n - 1)

        n = self.Num
        catalan = factorial(2 * n) // (factorial(n + 1) * factorial(n))
        return catalan

    def es_mersenne(self):
        p = (self.Num + 1).bit_length() - 1
        return self.es_primo(p) if (2**p - 1) == self.Num else False


def mostrar_menu():
    print("\nMenú de Opciones:")
    print("1. Ingresar número")
    print("2. Ver número actual")
    print("3. Incrementar número")
    print("4. Decrementar número")
    print("5. Verificar si es Par o Impar")
    print("6. Verificar si es Positivo o Negativo")
    print("7. Calcular Factorial")
    print("8. Verificar si es Número Perfecto")
    print("9. Verificar si es Primo")
    print("10. Mostrar sucesión Fibonacci")
    print("11. Verificar si es Número Armstrong")
    print("12. Verificar si es Perfecto Generalizado")
    print("13. Mostrar Fila de Pascal")
    print("14. Cifrar texto con César")
    print("15. Verificar si tiene Primo Gemelo")
    print("16. Descomponer en Factores Primos")
    print("17. Calcular Número de Catalan")
    print("18. Verificar si es Número de Mersenne")
    print("0. Salir")


def ejecutar_opcion(enter_number):
    while True:
        mostrar_menu()
        opcion = input("Seleccione una opción: ")

        if opcion == "1":
            numero = int(input("Ingrese un número: "))
            enter_number.set_num(numero)
        elif opcion == "2":
            print(f"El número actual es: {enter_number.get_num()}")
        elif opcion == "3":
            enter_number.incrementar_num()
            print(f"Número incrementado: {enter_number.get_num()}")
        elif opcion == "4":
            enter_number.decrementar_num()
            print(f"Número decrementado: {enter_number.get_num()}")
        elif opcion == "5":
            es_par = enter_number.es_par_impar()
            print("Es Par" if es_par else "Es Impar")
        elif opcion == "6":
            es_positivo = enter_number.es_negativo_positivo()
            print("Es Positivo" if es_positivo else "Es Negativo")
        elif opcion == "7":
            print(f"Factorial: {enter_number.factorial()}")
        elif opcion == "8":
            es_perfecto = enter_number.es_perfecto()
            print("Es Perfecto" if es_perfecto else "No es Perfecto")
        elif opcion == "9":
            es_primo = enter_number.es_primo()
            print("Es Primo" if es_primo else "No es Primo")
        elif opcion == "10":
            print(f"Fibonacci: {enter_number.fibonacci()}")
        elif opcion == "11":
            es_armstrong = enter_number.es_armstrong()
            print("Es Número Armstrong" if es_armstrong else "No es Número Armstrong")
        elif opcion == "12":
            k = int(input("Ingrese el valor de k: "))
            es_perfecto_gen = enter_number.es_perfecto_generalizado(k)
            print(f"Es Perfecto Generalizado con k={k}" if es_perfecto_gen else f"No es Perfecto Generalizado con k={k}")
        elif opcion == "13":
            print(f"Fila de Pascal: {enter_number.calcular_fila_pascal()}")
        elif opcion == "14":
            desplazamiento = int(input("Ingrese el valor del desplazamiento: "))
            texto = input("Ingrese el texto a cifrar: ")
            print(f"Cifrado César: {enter_number.cifrar_cesar(desplazamiento, texto)}")
        elif opcion == "15":
            tiene_gemelo = enter_number.tiene_primo_gemelo()
            print("Tiene Primo Gemelo" if tiene_gemelo else "No tiene Primo Gemelo")
        elif opcion == "16":
            print(f"Factores Primos: {enter_number.descomponer_en_factores_primos()}")
        elif opcion == "17":
            print(f"Número de Catalan: {enter_number.calcular_catalan()}")
        elif opcion == "18":
            es_mersenne = enter_number.es_mersenne()
            print("Es Número de Mersenne" if es_mersenne else "No es Número de Mersenne")
        elif opcion == "0":
            print("Saliendo del programa...")
            break
        else:
            print("Opción no válida, intente de nuevo.")


if __name__ == "__main__":
    numero = Entero()
    ejecutar_opcion(numero)

